### PR TITLE
[compiler] Removing copies of duplicate globals

### PR DESCRIPTION
**Pull request by NoumanAmir657**
## Identification of problem
While compiling the `llama-3.2-1B` model—using input MLIR generated by `shark-ai`. I observed that the model consumed approximately **5.8 GB** of host RAM, which is suboptimal for a model with **1 billion parameters** operating in `fp16`. The compilation targeted the **x86** backend with the `mmt4d` kernel enabled.

When `mmt4d` is used, both the left-hand side (LHS) and right-hand side (RHS) operands are **packed** prior to applying the `mmt4d` operation. In the input MLIR, the model defines both a `prefill` and a `decode` function. In each case, the same set of **constant weight tensors** are packed independently. Since these weights are constants, they are hoisted into their respective `util.initializer` regions. Consequently, both `prefill` and `decode` functions have their own `util.initializer`, each of which contains pack dispatches to prepare the weights.

The pass `iree-util-combine-initializers` subsequently merges the two initializers. However, this results in **duplicate pack dispatches**—that is, the same weight matrix is packed twice and stored in two distinct locations. Ideally, the IREE compiler should recognize and eliminate such redundancy when the packing parameters (e.g., tiling configuration) are identical. However, no subsequent pass in the IREE compilation pipeline deduplicates these pack operations, leading to approximately **2.5 GB** of unnecessary memory usage due to duplicated buffers.

## Smaller Test Case
```MLIR
module @module {
    util.global private @__lhs = #stream.parameter.named<"model"::"lhs"> : tensor<4096x4096xf32>

    func.func @m1(%arg0 : tensor<4096x4096xf32>) -> tensor<4096x4096xf32> {
        %cst = arith.constant 0.000000 : f32
        %__lhs = util.global.load @__lhs : tensor<4096x4096xf32>
        %10 = torch_c.from_builtin_tensor %__lhs : tensor<4096x4096xf32> -> !torch.vtensor<[4096,4096],f32>
        %11 = torch_c.to_builtin_tensor %10 : !torch.vtensor<[4096,4096],f32> -> tensor<4096x4096xf32>

        %12 = tensor.empty() : tensor<4096x4096xf32>
        %acc = linalg.fill ins(%cst : f32) outs(%12: tensor<4096x4096xf32>) -> tensor<4096x4096xf32>
        %result = linalg.matmul ins(%arg0, %11: tensor<4096x4096xf32>, tensor<4096x4096xf32>) outs(%acc: tensor<4096x4096xf32>) -> tensor<4096x4096xf32>
        return %result : tensor<4096x4096xf32>
    }

    func.func @m2(%arg0 : tensor<4096x4096xf32>) -> tensor<4096x4096xf32> {
        %cst = arith.constant 0.000000 : f32
        %__lhs = util.global.load @__lhs : tensor<4096x4096xf32>
        %10 = torch_c.from_builtin_tensor %__lhs : tensor<4096x4096xf32> -> !torch.vtensor<[4096,4096],f32>
        %11 = torch_c.to_builtin_tensor %10 : !torch.vtensor<[4096,4096],f32> -> tensor<4096x4096xf32>

        %12 = tensor.empty() : tensor<4096x4096xf32>
        %acc = linalg.fill ins(%cst : f32) outs(%12: tensor<4096x4096xf32>) -> tensor<4096x4096xf32>
        %result = linalg.matmul ins(%arg0, %11: tensor<4096x4096xf32>, tensor<4096x4096xf32>) outs(%acc: tensor<4096x4096xf32>) -> tensor<4096x4096xf32>
        return %result : tensor<4096x4096xf32>
    }
}
```

### Compilation command
```bash
iree-compile
--mlir-print-ir-after-all \
--iree-llvmcpu-target-cpu-features=host \
--iree-hal-target-device=local \
--iree-hal-local-target-device-backends=llvm-cpu \
in.mlir -o in.vmfb > in.dump 2>&1
```

### After `iree-util-combine-initializers` pass
```MLIR
// -----// IR Dump After CombineInitializersPass (iree-util-combine-initializers) //----- //
module @module attributes {stream.affinity.default = #hal.device.affinity<@__device_0>} {
  util.global private @__device_0 = #device_target_local
  util.global private @__lhs : !stream.resource<constant>
  util.global private @__lhs__size : index
  util.initializer {
    %cst = stream.tensor.constant on(#hal.device.affinity<@__device_0>) : tensor<4096x4096xf32> in !stream.resource<constant> = #stream.parameter.named<"model"::"lhs"> : tensor<4096x4096xf32>
    %0 = stream.resource.size %cst : !stream.resource<constant>
    util.global.store %cst, @__lhs : !stream.resource<constant>
    util.global.store %0, @__lhs__size : index
    %__lhs__size = util.global.load @__lhs__size : index
    %__lhs = util.global.load @__lhs : !stream.resource<constant>
    %1 = stream.async.transfer %__lhs : !stream.resource<constant>{%__lhs__size} from(#hal.device.affinity<@__device_0>) -> to(#hal.device.affinity<@__device_0>) !stream.resource<*>{%__lhs__size}
    %2 = stream.tensor.sizeof on(#hal.device.affinity<@__device_0>) tensor<512x4096x8x1xf32> : index
    %3 = stream.tensor.dispatch on(#hal.device.affinity<@__device_0>) @_initializer_0_dispatch_0::@_initializer_0_dispatch_0_pack_f32(%1) : (tensor<4096x4096xf32> in !stream.resource<*>{%__lhs__size}) -> tensor<512x4096x8x1xf32> in !stream.resource<*>{%2}
    %4 = stream.async.transfer %3 : !stream.resource<*>{%2} from(#hal.device.affinity<@__device_0>) -> to(#hal.device.affinity<@__device_0>) !stream.resource<constant>{%2}
    util.global.store %4, @__hoisted_tensor_512x4096x8x1xf32 : !stream.resource<constant>
    util.global.store %2, @__hoisted_tensor_512x4096x8x1xf32__size : index
    %__lhs__size_0 = util.global.load @__lhs__size : index
    %__lhs_1 = util.global.load @__lhs : !stream.resource<constant>
    %5 = stream.async.transfer %__lhs_1 : !stream.resource<constant>{%__lhs__size_0} from(#hal.device.affinity<@__device_0>) -> to(#hal.device.affinity<@__device_0>) !stream.resource<*>{%__lhs__size_0}
    %6 = stream.tensor.sizeof on(#hal.device.affinity<@__device_0>) tensor<512x4096x8x1xf32> : index
    %7 = stream.tensor.dispatch on(#hal.device.affinity<@__device_0>) @_initializer_0_dispatch_0::@_initializer_0_dispatch_0_pack_f32(%5) : (tensor<4096x4096xf32> in !stream.resource<*>{%__lhs__size_0}) -> tensor<512x4096x8x1xf32> in !stream.resource<*>{%6}
    %8 = stream.async.transfer %7 : !stream.resource<*>{%6} from(#hal.device.affinity<@__device_0>) -> to(#hal.device.affinity<@__device_0>) !stream.resource<constant>{%6}
    util.global.store %8, @__hoisted_tensor_512x4096x8x1xf32_0 : !stream.resource<constant>
    util.global.store %6, @__hoisted_tensor_512x4096x8x1xf32_0__size : index
    util.return
  }
  .
  .
  .
}
```
In the above, we can notice that `@__lhs` is getting packed twice in the `util.initializer` and stored in two different global variables. This same scenario above happens for the weights in `prefill` and `decode`.

## Implementation of Pass (old + not being used now + read Ben's comment below)
Currently, my implementation for removing these extra stores and dispatches is very naive but it works and I observed that the memory usage for llama-3.2-1B went down by 2.5 GB after this pass was added in the compilation flow

### Current implmentation
The pass `iree-util-optimize-global-duplicates` is run right after `iree-util-combine-initializers`. This happens in the `stream` compilation phase. The pass is not affinity aware as of now.
Currently, the pass looks for this pattern in the IR:
```MLIR
%__lhs = util.global.load @__lhs : !stream.resource<constant>
%1 = stream.async.transfer %__lhs : !stream.resource<constant>{%__lhs__size} from(#hal.device.affinity<@__device_0>) -> to(#hal.device.affinity<@__device_0>) !stream.resource<*>{%__lhs__size}
%2 = stream.tensor.sizeof on(#hal.device.affinity<@__device_0>) tensor<512x4096x8x1xf32> : index
%3 = stream.tensor.dispatch on(#hal.device.affinity<@__device_0>) @_initializer_0_dispatch_0::@_initializer_0_dispatch_0_pack_f32(%1) : (tensor<4096x4096xf32> in !stream.resource<*>{%__lhs__size}) -> tensor<512x4096x8x1xf32> in !stream.resource<*>{%2}
%4 = stream.async.transfer %3 : !stream.resource<*>{%2} from(#hal.device.affinity<@__device_0>) -> to(#hal.device.affinity<@__device_0>) !stream.resource<constant>{%2}
util.global.store %4, @__hoisted_tensor_512x4096x8x1xf32 : !stream.resource<constant>
```

A map is maintained where each key corresponds to a global variable, such as `@__lhs`. The value associated with each key is a pair consisting of:
- a `SymbolRefAttr`, which refers to the dispatch operation responsible for packing the weight, and
- a `StringRef`, which names the global variable where the result of that packing (i.e., the packed weight tensor) is stored—for example, `@__hoisted_tensor_512x4096x8x1xf32`.

When another instance of the same pattern is encountered—i.e., one that packs the same weight tensor using the same tiling configuration—we treat it as redundant. In such cases, the corresponding `util.global.store` operation is removed, which causes the dispatch to become unused and allows it to be eliminated by subsequent compiler passes. All uses of the removed store are then updated to reference the previously hoisted global, thereby avoiding duplication.

Would appreciate some form of initial review for this to know if I am moving in the right direction!

### CODE REVIEW

**Review by benvanik — CHANGES_REQUESTED** [review]
Stepping back, this approach will not work and is duplicative (hah!) with passes we already have. Specifically, the FoldGlobalsPass and FuseGlobalsPass should already be identifying globals that have the same value while ignoring affinity. At layers that are not affinity-aware (so before stream) those passes are what should be used. They already run at the tail of the flow pipeline and head of the stream pipeline when it's safer to do this.

To better handle hoisted initializers then FoldGlobalsPass/FuseGlobalsPass will have to get smarter about checking initialization slices instead of specific initialization values as they do today. If two globals have the same exact side-effect-free tree of ops producing their value they can be folded. FuseGlobalsPass today just checks whether the same "value" is stored but it could be checking for structurally equivalent side-effect-free slices. Probably something around `initialValueMap` which could instead be walking up all the use-def chains of producing the value and comparing for structural equivalence and memory effects.

As a general review comment, this pass is violating layering by being in Util/ (a generic utility dialect used by all other dialects/all phases of compilation) but referencing Stream/ code/ops. It cannot do that - nor even reason about what a "dispatch" is. Folding globals produced by equivalent side-effect-free slices does not require any of that coupling, anyway.

As for affinity, anything happening on stream ops must be affinity-aware. This specific problem does not require that, though, as it should be happening prior to the stream dialect.

**Comment by NoumanAmir657** [comment]
Thanks @benvanik 

The pass coupling with dispatches and stream was an oversight on my part.
From what you said above, we should just tackle this problem in the FoldGlobalPass or FuseGlobalPass instead of introducing a completely new pass for this.

**Comment by NoumanAmir657** [comment]
@benvanik I went over the code of `FuseGlobalsPass` and as you mentioned before, that pass only works for same SSA values getting assigned to some global.
In order to analyze the tree of ops that lead to that store in global, I looked up `mlir::OperationEquivalence` struct. Comparing the structural equivalence of ops using `mlir::OperationEquivalence`, and then making the fusable sets can be one way we can go about this.

Would like to know what you think about this. And I apologize if my questions or approach are too naive. This is my first time contributing to MLIR/IREE.

**Comment by NoumanAmir657** [comment]
After investigating a bit more, I found this `isStructurallyEquivalentTo` in `compiler/src/iree/compiler/Utils/EquivalenceUtils.h`. It seems that currently it does not support `SymbolRefAttrs`. 
Related to [this](https://github.com/iree-org/iree/issues/3996)

**Comment by benvanik** [comment]
isStructurallyEquivalentTo with the cache is what you'll want to use. Currently it has the #3996 TODO about symbols that would make it not work for this, but that could be fixed by adding an optional `std::function<bool(SymbolRefAttr, SymbolRefAttr)> areSymbolsEquivalent` arg to the function. By default it could return true only if the attributes are equal and otherwise and existing call sites could pass in one that always returned true, e.g.
```C++
    if (cache.isSymbolAttrName(lhsAttr.getName())) {
      if (!areSymbolsEquivalent(lhsAttr, rhsAttr))
        return false;
    } else {
      if (lhsAttr != rhsAttr)
        return false;
    }
```

Thinking about ways to keep this simple, a `CSEInitializersPass` in Util may be relatively simple and allow the Fuse/FoldGlobalsPasses to remain the same. It'd gather all initializers in the program that have side-effect-free slices terminating in global stores, build `EquivalenceClasses` using the equivalence routine, and then replace all initializers in each class with one that has a clone of the slice but storing to all the original globals. `FuseGlobalsPass` would then trivially find those and fuse the globals as a separate (much easier to debug) pass.

e.g.
```mlir
util.global @global_a
util.initializer {
   %... = some.ops.unrelated (not in slice, must be preserved)
   %0 = produce
   %1 = mutate %0 
   util.global.store %1, @global_a
   util.return
}
//
util.global @other_globals...
util.initializer { other initialization }
//
util.global @global_b
util.initializer {
   %... = some.ops.unrelated (not in slice, must be preserved)
   %0 = produce
   %1 = mutate %0 
   util.global.store %1, @global_b
   util.return
}
//
util.global @other_globals...
util.initializer { other initialization }
```
->
```mlir
util.global immutable @global_a
// global_b moved up to the lead global in an equivalence class (preserving program order)
util.global immutable @global_b
// original initializer for @global_a is preserved
util.initializer {
   %... = some.ops.unrelated (not in slice, must be preserved)
   %0 = produce
   %1 = mutate %0 
   // original store of @global_a preserved
   util.global.store %1, @global_a
   // new store to @global_b inserted using the same SSA value as @global_a
   util.global.store %1, @global_b
   util.return
}
//
util.global @other_globals...
util.initializer { other initialization }
//
util.initializer {
   %... = some.ops.unrelated (not in slice, must be preserved)
   %0 = produce
   %1 = mutate %0 
   // @global_b store is erased - the rest will be erased by DCE (if able)
   util.return
}
//
util.global @other_globals...
util.initializer { other initialization }
```

The tricky part here is preserving program correctness: initializers must only reference globals declared earlier in module order aka globals must be declared before they are used in module order. Since this pass would only be touching globals that are known to be safe to CSE the global and its initialization can be hoisted up to the first global in the equivalence class. This allows subsequent initializers that may have been using the global to still reference it in its initialized state.

I think this would be easy to write/debug in isolation as all it would be doing is moving an immutable global earlier in program order (always safe) and moving the initialization of that global earlier in initialization order (always safe for immutable ones). This should work if those slices originate as immutable global loads from parameters (I believe the case you're interested in, where parameters are loaded and then packed on startup) as the global load is side-effect free and the lead global in the equivalence class would by definition have a valid claim to using the parameter global and we aren't reordering past that (the lead global/it's initializer never change besides the new store ops). Mutable globals would not be safe to CSE without a lot more analysis, but parameter globals are always immutable (today) so it should be fine.

Does that make sense? I'm designing off the cuff :)

**Comment by NoumanAmir657** [comment]
From what we want to do in this pass, it seems the best place to run this pass is exactly after `util.initializers` are materialized. All the initializers that loads an immutable global (parameter) -> some op on that parameter -> store will be separate and we can iterate over all initializers and gather the ones that have same immutable load together and then use  `IsStructurallyEquivalent` on the initializer region. And then we use equivalence classes routine

One question that I have is that
```MLIR
util.initializer { 
    %... = some.ops.unrelated (not in slice, must be preserved) 
    %0 = produce 
    %1 = mutate %0 
    util.global.store %1, @global_a
    util.return 
}
```
Can `some.ops.unrelated` slice exist the first time `util.initializer` is materialized. I am assuming independent slices will be in a separate initializer at first.

**Comment by benvanik** [comment]
Responding to your question then will take a look at the new code!

> From what we want to do in this pass, it seems the best place to run this pass is exactly after util.initializers are materialized.

> Can `some.ops.unrelated` slice exist the first time `util.initializer` is materialized. I am assuming independent slices will be in a separate initializer at first.

We can't assume anything about initializers as they are part of program input: a frontend can make them do whatever they want. We never distinguish between initializers (or functions) we create vs ones that are input - the only thing we can control is how we create initializers. We also may create them many different ways at many different places - for example, the torch import process may turn special "load from file" ops into initialized globals with file I/O operations before we get to parameter handling, or a custom compiler pipeline may insert custom module logic, etc. Basically, the pass must always be safe to run at whatever phase of the compiler it can run in, and when it comes to things manipulating core ops (like initializers/functions/etc) they've got to be very safe (or quick to bail if they can't be safe).

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp [hunk 1]
```diff
@@ -0,0 +1,137 @@
+// Copyright 2021 The IREE Authors
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: 2025 (always use the year of file creation)

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+
+#define DEBUG_TYPE "iree-util-cse-initializer"
+
+namespace mlir::iree_compiler::IREE::Util {
+
+#define GEN_PASS_DEF_CSEINITIALIZERSPASS
+#include "iree/compiler/Dialect/Util/Transforms/Passes.h.inc"
+
+namespace {
+static llvm::cl::opt<bool> clDisableCSEInitializers(
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: a few things:

- do not name flags with "enable" or "disable" - a boolean flag already explicitly indicates that: if "--iree-util-cse-initializers=true" then it's enabled, and if false then it's not

- do not have flags that control pass behavior in passes - such options are part of the pipeline that decides whether to add the pass or not

- any pipeline that runs this pass should have its own option for whether to cse, if it's even needed

- something as fundamental as cse like this must be safe, and must be always profitable - there shouldn't need to be a flag to disable it - if it does have cost model/heuristic behavior that may cause a user to want to disable it then those should be exposed per-pipeline and per-heuristic ("only cse initializers of scalar values" or "only cse initializers that only store a single global" or "only cse initializers that do not contain any other ops but those that can be csed" etc) - but those are best to add only when required and no sooner (every flag/option is a maintenance and testing burden and must pay for itself)

so, here, remove this flag

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+    "iree-disable-util-cse-initializers",
+    llvm::cl::desc("Disables the iree-util-cse-initializers pass"),
+    llvm::cl::init(false));
+
+bool areAttributesEquivalent(Operation &lhs, Operation &rhs) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]:
```suggestion
static bool areAttributesEquivalent(Operation &lhs, Operation &rhs) {
```

always put `static` on functions that are private to a compilation unit (anonymous namespace does the same thing but is less readable - a reader skimming the file has to scroll up and detect if they are in an anonymous region, as code moves it may get moved out of an anonymous region accidentally, etc)

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+  auto storeOpLhs = llvm::dyn_cast_or_null<IREE::Util::GlobalStoreOp>(lhs);
+  auto storeOpRhs = llvm::dyn_cast_or_null<IREE::Util::GlobalStoreOp>(rhs);
+  if (storeOpLhs && storeOpRhs) {
+    return true;
+  }
+  return false;
+}
+class CSEInitializersPass
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp:31]:
```suggestion

class CSEInitializersPass
```

File: compiler/src/iree/compiler/Utils/EquivalenceUtils.cpp [hunk 5]
```diff
@@ -209,38 +225,43 @@ static bool isStructurallyEquivalentTo(OperationEquivalenceCache &cache,
...
 
-  // TODO(#3996): symbol mapping; for now allow them to differ unconditionally.
+  // TODO(#3996): symbol mapping; for now allow them to differ
+  // unconditionally.
   if (lhsEntry.attrs.getAttrs().size() != rhsEntry.attrs.getAttrs().size())
     return false;
   for (auto [lhsAttr, rhsAttr] :
        llvm::zip_equal(lhsEntry.attrs, rhsEntry.attrs)) {
     if (!cache.isSymbolAttrName(lhsAttr.getName())) {
+      if (areAttributesEquivalent && areAttributesEquivalent(lhs, rhs)) {
```
Comment (benvanik) [inline compiler/src/iree/compiler/Utils/EquivalenceUtils.cpp]: As mentioned in https://github.com/iree-org/iree/pull/21130#issuecomment-3009238402 this must be per-attribute not per op - this only needs to check that the lhsAttr symbol ref attr is equivalent to the rhsAttr symbol ref attr - it does not need to take the op. `areAttributesEquivalent` implies the caller must check whether _all_ attributes on the ops are equivalent - which may be useful in some other cases but is not what's implemented here (this is only calling on the sym ref attrs and only checking specific ops). The casting performed in your implementation in CSEInitializers is also not safe/correct as it's allowing any global stores to be considered equivalent when that's not what you want. If the callback says two ops attributes are not equivalent you still check below and that could return true, making the callback not do what it is supposed to.

This should be taking only sym ref attrs as `areSymbolsEquivalent` as in the pseudo code in the original comment, and the callback you have should never be called on a global store as you should never be trying to compare those, only the ops in the slice leading up to the store (so global loads/etc).

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp [hunk 6]
```diff
@@ -0,0 +1,137 @@
...
+  }
+
+  void runOnOperation() override {
+    if (clDisableCSEInitializers)
+      return;
+
+    auto moduleOp = getOperation();
+
+    SmallVector<IREE::Util::InitializerOp> initOps;
+    llvm::SmallDenseMap<IREE::Util::InitializerOp, IREE::Util::GlobalStoreOp>
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: always use `GlobalOpInterface`/`GlobalLoadOpInterface`/`GlobalStoreOpInterface` instead of the explicit global ops. This allows other global ops in other dialects that may be present to still be handled. As a general rule if an interface exists on an op that can satisfy what you need you should always use it even if _today_ you only need it for one concrete op.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp [hunk 7]
```diff
@@ -0,0 +1,137 @@
...
+    llvm::EquivalenceClasses<IREE::Util::InitializerOp> ec;
+    for (int i = 0; i < initOps.size(); ++i) {
+      if (initOpBitVector[i]) {
+        continue; // Already processed.
+      }
+      for (int j = i + 1; j < initOps.size(); ++j) {
+        if (initOpBitVector[j]) {
+          continue; // Already processed.
+        }
+        if (isStructurallyEquivalentTo(
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: This needs to be per-op in a slice, not per-initializer. As mentioned in https://github.com/iree-org/iree/pull/21130#issuecomment-3009238402 there may be many other ops in an initializer that don't matter to the particular global being initialized _and_ there may be multiple globals initialized by the same ops. This pass needs to look at the ops that initialize a global and deduplicate those expressions.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp [hunk 8]
```diff
@@ -0,0 +1,137 @@
...
+      bool allImmutableLoads = true;
+      for (auto loadOp : callableOp.getInitializerRegion()
+                             .getOps<IREE::Util::GlobalLoadOp>()) {
+        if (!loadOp.isGlobalImmutable()) {
+          allImmutableLoads = false;
+          break;
+        }
+      }
+
+      auto bg = callableOp.getInitializerRegion()
```
Comment (benvanik) [inline compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: This kind of logic on iterators is tricky to follow and generally should be avoided - if you need to check the count of something then `llvm::size` can be used instead. But as mentioned elsewhere this is not the kind of logic the pass should be doing (it should be working back from global stores to the slices feeding into them).

**Review by benvanik — CHANGES_REQUESTED** [review]
This will need some more cooking. Please check https://github.com/iree-org/iree/pull/21130#issuecomment-3009238402 again for the flow. You can't assume there's always a global store, or that there's only one global stored per initializer, or that a single expression doesn't get stored to multiple globals. Instead of thinking about this as initializers (the `util.initializer` op) you are CSEing you're CSEing the initialization of globals. 

So:
1. Find all globals
1. Find the stores into those globals
1. Build the per-global slices backwards to find side-effect free ops
1. (Maybe) grow the slices forward to find if those slices store to other globals
1. Build the equivalence classes based on those slices (not their initializers); slices that store to multiple globals are equivalent by construction
1. As mentioned in https://github.com/iree-org/iree/pull/21130#issuecomment-3009238402 you need to move globals to preserve initialization order

The first two (finding globals and their stores) is provided by `compiler/src/iree/compiler/Dialect/Util/Analysis/GlobalTable.h` - that gives you what you need to be safe: for each global you can ignore indirect ones (`isIndirect == true`), only inspect those stored exclusively from initializers (`onlyInitialized == true`), and then get all the stores into the global (`storeOps`). If there's more than one store op you can skip the global as that's a more complex situation.

`mlir::getSlice` is what you want to use for the slices as it can walk backward and forward to find the full set of operations involved. You want it to stop the backward walk on mutable global loads and ops with side effects at the least, but possibly others. You want it to stop the forward walk on side-effecting ops but include global stores so that you can handle initializers with multiple globals. The forward walk may not be required and the same behavior may arise naturally from just doing slice equivalence - best decided with tests/inspection.

The target initializer chosen from an equivalence class to be the one that stores to all equivalent globals must be the first one in module order, and all of the globals stored must be moved above it. This is because initializers are run in module order and globals must be defined before they are initialized.

You'll also need a lot of tests for the pass both positive (CSE'd), negative (not CSE'd), and mixed (some globals in an initializer CSE'd while others not.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp (hunk 8 above)
Comment (NoumanAmir657) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: Fixed.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp (hunk 7 above)
Comment (NoumanAmir657) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: This is fixed. The pass now uses `mlir::slice` utility to gather a backward slice that stops when it find a side effecting op or when it encounters a mutable load.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp (hunk 6 above)
Comment (NoumanAmir657) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: Fixed.

File: compiler/src/iree/compiler/Utils/EquivalenceUtils.cpp (hunk 5 above)
Comment (NoumanAmir657) [reply compiler/src/iree/compiler/Utils/EquivalenceUtils.cpp]: I think I lack some basic understanding for `SymbolRefAttr`. I don't understand why we need a separate logic for it. Keeping the logic same as it is makes the pass work. Can you tell what I may be missing.
In the new commit, I don't compare the stores, only the slice leading up to the store. I am using the `Operation` overloaded version of the `isStructurallyEquivalent` with cache now.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp (hunk 3 above)
Comment (NoumanAmir657) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: This is removed now.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp (hunk 2 above)
Comment (NoumanAmir657) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: Removed.

File: compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp (hunk 1 above)
Comment (NoumanAmir657) [reply compiler/src/iree/compiler/Dialect/Util/Transforms/CSEInitializers.cpp]: Fixed.

**Comment by NoumanAmir657** [comment]
@benvanik In the pass, do we need to move the globals declaration like as you mentioned in this:
```MLIR
util.global @global_a
util.initializer {
   %... = some.ops.unrelated (not in slice, must be preserved)
   %0 = produce
   %1 = mutate %0 
   util.global.store %1, @global_a
   util.return
}
//
util.global @other_globals...
util.initializer { other initialization }
//
util.global @global_b
util.initializer {
   %... = some.ops.unrelated (not in slice, must be preserved)
   %0 = produce
   %1 = mutate %0 
   util.global.store %1, @global_b
   util.return
}
//
util.global @other_globals...
util.initializer { other initialization }
```
->
```MLIR
util.global immutable @global_a
// global_b moved up to the lead global in an equivalence class (preserving program order)
util.global immutable @global_b
// original initializer for @global_a is preserved
util.initializer {
   %... = some.ops.unrelated (not in slice, must be preserved)
   %0 = produce
   %1 = mutate %0 
   // original store of @global_a preserved
   util.global.store %1, @global_a
   // new store to @global_b inserted using the same SSA value as @global_a
   util.global.store %1, @global_b
   util.return
}
```
Since we are counting on the `fuse-globals` pass to fuse these globals later on. Then would it matter if we moved the `
util.global immutable @global_b` above?
Although not moving them does make the IR illegal I think so we should do it.

**Edit:** I moved them in the latest commit.

**Comment by NoumanAmir657** [comment]
@benvanik 
Can you review this again! Thanks!

**Comment by hanhanW** [comment]
I only looked at the PR description, and it should be addressed by https://github.com/iree-org/iree/pull/21921

The epic issue was https://github.com/iree-org/iree/issues/21659, and I've been working 2x -> 1x memory footprint issue, tracking issue: https://github.com/iree-org/iree/issues/22371

Feel free to reopen it if there are remaining action items.
